controllers/metal3.io/bmceventsubscription_controller.go

```diff
@@ -0,0 +1,260 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	metal3v1alpha1 "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+	"github.com/metal3-io/baremetal-operator/pkg/ironic/bmc"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner"
+	"github.com/metal3-io/baremetal-operator/pkg/utils"
+)
+
+const (
+	subscriptionRetryDelay = time.Minute * 10
+)
+
+type BMCEventSubscriptionReconciler struct {
+	client.Client
+	Log                logr.Logger
+	ProvisionerFactory provisioner.Factory
+}
+
+func (r *BMCEventSubscriptionReconciler) Reconcile(ctx context.Context, request ctrl.Request) (result ctrl.Result, err error) {
+	reqLogger := r.Log.WithValues("bmceventsubscription", request.NamespacedName)
+	reqLogger.Info("start")
+
+	defer func() {
+		reqLogger.Info("done")
+	}()
+
+	// Fetch the BMCEventSubscription
+	subscription := &metal3v1alpha1.BMCEventSubscription{}
+	err = r.Get(ctx, request.NamespacedName, subscription)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after
+			// reconcile request.  Owned objects are automatically
+			// garbage collected. For additional cleanup logic use
+			// finalizers.  Return and don't requeue
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, errors.Wrap(err, "could not load subscription")
+	}
+
+	host := &metal3v1alpha1.BareMetalHost{}
+	namespacedHostName := types.NamespacedName{
+		Name:      subscription.Spec.HostName,
+		Namespace: request.Namespace,
+	}
+	err = r.Get(ctx, namespacedHostName, host)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			reqLogger.Error(err, "baremetalhost not found")
+			return ctrl.Result{RequeueAfter: subscriptionRetryDelay}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, errors.Wrap(err, "could not load host data")
+	}
+
+	// Add a finalizer to newly created objects.
+	if subscription.DeletionTimestamp.IsZero() && !subscriptionHasFinalizer(subscription) {
+		reqLogger.Info(
+			"adding finalizer",
+			"existingFinalizers", subscription.Finalizers,
+			"newValue", metal3v1alpha1.BMCEventSubscriptionFinalizer,
+		)
+		subscription.Finalizers = append(subscription.Finalizers,
+			metal3v1alpha1.BMCEventSubscriptionFinalizer)
+		err := r.Update(ctx, subscription)
+		if err != nil {
+			return ctrl.Result{}, errors.Wrap(err, "failed to add finalizer")
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	// Retrieve the BMC details from the host spec and validate host
+	// BMC details and build the credentials for talking to the
+	// management controller.
+	var bmcCreds *bmc.Credentials
+
+	bmcCreds, _, err = r.buildAndValidateBMCCredentials(request, host)
+	if err != nil || bmcCreds == nil {
```

Comment:
We shouldn't return bmcCreds == nil without also returning an err, so the 2nd check seems redundant